core/templates/dev/head/pages/exploration_editor/simple_editor_tab/SimpleEditorBodyDirective.js

```diff
@@ -0,0 +1,128 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Directive for the body of the simple editor.
+ * editor.
+ */
+
+oppia.directive('simpleEditorBody', [function() {
+  return {
+    restrict: 'E',
+    scope: {},
+    templateUrl: 'simpleEditor/body',
+    controller: [
+      '$scope', '$rootScope', '$timeout', 'SimpleEditorQuestionsDataService',
+      'explorationTitleService', 'explorationInitStateNameService',
+      'explorationStatesService',
+      function(
+          $scope, $rootScope, $timeout, SimpleEditorQuestionsDataService,
+          explorationTitleService, explorationInitStateNameService,
+          explorationStatesService) {
+        $scope.explorationTitleService = explorationTitleService;
+        $scope.questions = [];
+
+        $scope.$on('simpleEditorLoaded', function() {
+          $scope.questions = (
+              SimpleEditorQuestionsDataService.getQuestionsInOrder());
+          $scope.initialState = explorationStatesService.getState(
+            explorationInitStateNameService.savedMemento);
+        });
+
+        $scope.saveTitle = function(newTitle) {
+          explorationTitleService.displayed = newTitle;
+          explorationTitleService.saveDisplayedValue();
+        };
+        $scope.saveIntroduction = function(newHtml) {
+          var initStateName = explorationInitStateNameService.savedMemento;
+          explorationStatesService.saveStateContent(initStateName, [{
+            type: 'text',
+            value: newHtml
+          }]);
+          // We reload the initialState because a fresh copy of the state is
+          // created in the states dict on every update.
+          $scope.initialState = explorationStatesService.getState(
+            explorationInitStateNameService.savedMemento);
+        };
+
+        $scope.saveCustomizationArgs = function(
+            stateName, newCustomizationArgs) {
+          explorationStatesService.saveInteractionCustomizationArgs(
+            stateName, newCustomizationArgs);
+        };
+        $scope.saveAnswerGroups = function(stateName, newAnswerGroups) {
+          explorationStatesService.saveInteractionAnswerGroups(
+            stateName, newAnswerGroups);
+        };
+        $scope.saveDefaultOutcome = function(stateName, newDefaultOutcome) {
+          explorationStatesService.saveInteractionDefaultOutcome(
+            stateName, newDefaultOutcome);
+        };
+        $scope.saveBridgeText = function(stateName, newHtml) {
+          explorationStatesService.saveStateContent(stateName, [{
+            type: 'text',
+            value: newHtml
+          }]);
+          // We reload the questions because a fresh copy of the state is
+          // created in the states dict on every update.
+          // TODO(sll): Pass this back to the questions data service? But we
```

Comment:
Either the questions need to be rebuilt from scratch, or perhaps an incremental update is possible. ``CollectionLinearizerService`` takes the second approach; it will initially construct the linear state from an arbitrary collection structure, then subsequent changes are cheap because they just modify that structure.